receiver/sqlqueryreceiver/scraper.go

```diff
@@ -0,0 +1,83 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sqlqueryreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/sqlqueryreceiver"
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+type scraper struct {
+	id                 config.ComponentID
+	query              Query
+	clientProviderFunc clientProviderFunc
+	dbProviderFunc     dbProviderFunc
+	logger             *zap.Logger
+	client             dbClient
+	db                 *sql.DB
+}
+
+var _ scraperhelper.Scraper = (*scraper)(nil)
+
+func (s scraper) ID() config.ComponentID {
+	return s.id
+}
+
+func (s *scraper) Start(context.Context, component.Host) error {
+	var err error
+	s.db, err = s.dbProviderFunc()
+	if err != nil {
+		return fmt.Errorf("failed to open db connection: %w", err)
+	}
+	s.client = s.clientProviderFunc(s.db, s.query.SQL, s.logger)
+	return nil
+}
+
+func (s scraper) Scrape(ctx context.Context) (pmetric.Metrics, error) {
+	out := pmetric.NewMetrics()
+	rows, err := s.client.metricRows(ctx)
+	if err != nil {
+		return out, fmt.Errorf("scraper: %w", err)
+	}
+	rms := out.ResourceMetrics()
+	rm := rms.AppendEmpty()
+	sms := rm.ScopeMetrics()
+	sm := sms.AppendEmpty()
+	ms := sm.Metrics()
+	var errs error
+	for _, metricCfg := range s.query.Metrics {
+		for _, row := range rows {
+			if err = rowToMetric(row, metricCfg, ms.AppendEmpty()); err != nil {
+				err = multierr.Append(errs, err)
+			}
+		}
+	}
+	if err != nil {
+		errs = fmt.Errorf("scraper.Scrape errors: %w", errs)
+	}
```

Comment:
Sorry for the late comment here, but I ran into an issue with this while doing some of my own testing. Currently we're merging `errs` and `err` into the `err` variable on line 71, repeatedly, for all errors we encounter. However, we're re-setting `err` every loop, meaning it'll only hold one error.

This means once we get to the check `if err != nil` we'll get into the if statement with just the last error. Then, in the `fmt.Errorf` statement we're using the `errs` variable, which hasn't actually been set at this point.

The current error message:
```
2022-06-23T15:14:23.606-0700	error	scraperhelper/scrapercontroller.go:197	Error scraping metrics	{"kind": "receiver", "name": "sqlquery", "error": "scraper.Scrape errors: %!w(<nil>)", "scraper": "sqlqueryreceiver/query-0: select count(*) as count, genre from sys.movie group by genre"}
```

I believe the solution here is to do `errs = multierr.Append(errs, err)` and then check `if errs != nil` below.

Reply:
oops, will fix.

Reply:
Done.

(I hadn't actually tested the contents of the errors so missed this dumb mistake. Added a test and made existing tests more explicit as well.)